docs(gatsby-plugin-sharp): Base64 features

polarathene opened the pull request:
## Description

Documents some base64 features that Gatsby has [supported for a while now](https://github.com/gatsbyjs/gatsby/pull/10944). These weren't easily discoverable or clear how to use them.

I did some minor refactoring related to these features. The relevant individual commits provide details in their commit messages.

### Documentation

The graphql parameters `toFormatBase64` and `base64Width` are now documented in the `gatsby-plugin-sharp` README. The global override equivalent options are mentioned for each, but not documented elsewhere.

@gatsbyjs/learning is there any additional documentation of these that would be appreciated? A code snippet or PR to a project that demonstrates the minor features?

### Improvements

Some code refactoring, fixes and tests added.

Also `background` option is now passed to `base64()`, fixes #25640

wardpeet commented on packages/gatsby-plugin-sharp/README.md:
```diff
@@ -30,9 +30,13 @@ plugins: [
+      useMozJpeg: process.env.GATSBY_JPEG_ENCODER === `MOZJPEG`,
       stripMetadata: true,
-      defaultQuality: 75,
+      lazyImageGeneration: true,
```
Currently, this argument doesn't do anything. So we shouldn't display it in my opinion.
```suggestion
```
```

LekoArts commented:
Seems like tests are failing cc @wardpeet @polarathene

polarathene commented:
**build-test ** failure seems unrelated?

Unit tests is a mock snapshot that needs to be updated, I'll take care of that.

polarathene commented:
I'm not familiar with the testing code. When sharp pipeline calls `resize()`, many tests are failing with the error:

```
Expected positive integer for height but received NaN of type number
```

I am assuming this is because of `base64Width` being moved out from a function call(`defaultBase64Width()`) which provided the default value, to now depending on the default value from `pluginDefaults` object which now lives in `plugin-options.js`. Should the tests be including those defaults?

polarathene commented:
I'm unable to get a working test environment locally for `gatsby-plugin-sharp`, jest raises failures about finding modules that should be available but it can't find them.

If someone has time to assist that would be helpful.Otherwise not much else I can do with this PR.

polarathene commented:
All tests should be passing, "Cloud Tests" seems to be hanging for some reason.

polarathene commented:
Added a fix for `background` option not being passed from fluid/fixed methods to the base64 method.

Included a basic test , although like the base64 tests I've added I'm not sure if they're sufficient, should I also be testing these via fluid/fixed tests instead of directly on base64? The bug for `background` was simply that the option wasn't passed on to the method, is it worth adding tests for that? 

I'm not sure if I can mock the base64 method called in that test file for a single test that ensures supported options are passed through?

polarathene commented:
@wardpeet Could you review/merge please?

polarathene commented:
@wardpeet @pvdz hello, friendly reminder request to take the next steps in resolving this PR, thank you.

muescha commented on packages/gatsby-plugin-sharp/README.md:
```diff
@@ -147,6 +152,26 @@ following:
 
+#### toFormatBase64
+
+base64 image uses the image format from the source , or the value of `toFormat`. This setting allows a different image format instead, available options are: `JPG`, `PNG`, `WEBP`.
```
in code they are lowercased and there is no case insensitive comparison is used

polarathene replied:
It uses uppercase like the docs above it for `toFormat`. It calls `healOptions()` to lowercase just like `toFormat` does.

muescha commented on packages/gatsby-plugin-sharp/src/index.js:
```diff
@@ -323,16 +323,16 @@ async function generateBase64({ file, args = {}, reporter }) {
       compressionLevel: options.pngCompressionLevel,
       adaptiveFiltering: false,
-      force: args.toFormat === `png`,
+      force: options.toFormat === `png`,
```
better compare case insensitive?

polarathene replied:
It's handled earlier in the `healOptions()` call:

https://github.com/gatsbyjs/gatsby/blob/1419afe21a48fc4d75909e054760693fadc8675f/packages/gatsby-plugin-sharp/src/plugin-options.js#L83-L84

polarathene commented:
> Added one question, looks good besides that.

@wardpeet where is that question?

wardpeet commented on packages/gatsby-plugin-sharp/src/plugin-options.js:
```diff
@@ -2,7 +2,8 @@ const _ = require(`lodash`)
 const pluginDefaults = {
-  forceBase64Format: false,
+  base64Width: 20,
+  forceBase64Format: ``, // valid formats: png,jpg,webp
```
This got omitted somehow, I probably didn't click on start-review...
Can't we change this to undefined instead? Technically this becomes an enum with 3 values then. We could add validation later on.
```suggestion
```

polarathene replied:
@wardpeet You want to change the value of empty string to `undefined`? I've just matched the value to what is already used for `toFormat`

https://github.com/gatsbyjs/gatsby/blob/1419afe21a48fc4d75909e054760693fadc8675f/packages/gatsby-plugin-sharp/README.md#L142-L143

https://github.com/gatsbyjs/gatsby/blob/1419afe21a48fc4d75909e054760693fadc8675f/packages/gatsby-plugin-sharp/src/plugin-options.js#L26-L27

`toFormatBase64` like `toFormat` is the per image option, while `forceBase64Format` is a global default setting (perhaps the name doesn't communicate that well? It's in `pluginDefaults` object, so should hopefully be clear that it's forcing the default format for base64?)

wardpeet replied:
That's what I was missing :) Thank you for the explanation! 👍 🌮
